Fixup: VM with multiple interfaces boot failure

sathnaga opened the pull request:
When VM have multiple interfaces and
when IP gets assigned to random interface,
get_address hang on single interface and timeout for boot(import)test results to failure.

This patch will address it by introducing
a parameter""flexible_nic_index"",
when set it will try to poke all the interface and
check which has valid IP and connect to it.

Fixes: https://github.com/avocado-framework/avocado-vt/issues/617

sathnaga commented:
TESTS      : 4
  (1/4) guest_sanity.boundary.network_max.spaprvlan.qemu.qcow2.scsi.smp2.spapr-vlan.CentOS.7.2.ppc64le.io-github-autotest-qemu.unattended_install.import.import.default_install.aio_native: PASS (43.07 s)
  (2/4) guest_sanity.boundary.network_max.spaprvlan.qemu.qcow2.scsi.smp2.spapr-vlan.CentOS.7.2.ppc64le.io-github-autotest-libvirt.remove_guest.without_disk: PASS (1.28 s)
  (3/4) guest_sanity.boundary.network_max.virtionet.qemu.qcow2.scsi.smp2.virtio_net.CentOS.7.2.ppc64le.io-github-autotest-qemu.unattended_install.import.import.default_install.aio_native: PASS (99.53 s)
  (4/4) guest_sanity.boundary.network_max.virtionet.qemu.qcow2.scsi.smp2.virtio_net.CentOS.7.2.ppc64le.io-github-autotest-libvirt.remove_guest.without_disk: PASS (1.33 s)
 RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0


 TESTS      : 4
  (1/4) guest_sanity.boundary.network_max.spaprvlan.qemu.qcow2.scsi.smp2.spapr-vlan.Fedora.24.ppc64le.io-github-autotest-qemu.unattended_install.import.import.default_install.aio_native: PASS (47.83 s)
  (2/4) guest_sanity.boundary.network_max.spaprvlan.qemu.qcow2.scsi.smp2.spapr-vlan.Fedora.24.ppc64le.io-github-autotest-libvirt.remove_guest.without_disk: PASS (1.29 s)
  (3/4) guest_sanity.boundary.network_max.virtionet.qemu.qcow2.scsi.smp2.virtio_net.Fedora.24.ppc64le.io-github-autotest-qemu.unattended_install.import.import.default_install.aio_native: \
 PASS (92.73 s)
  (4/4)

sathnaga commented:
Used Test config:
```
include tests-shared.cfg
username = root
password = 123456
main_vm = virt-tests-vm1
vms = virt-tests-vm1
#Network
nettype = bridge
netdst=virbr0
# Using Text mode
display = 'nographic'
take_regular_screendumps = no
keep_screendumps_on_error = no
keep_screendumps = no
store_vm_register = no
virt_install_binary = /usr/bin/virt-install
qemu_img_binary = /usr/bin/qemu-img
hvm_or_pv = hvm
machine_type = pseries
only bridge
no xen, lxc, esx, ovmf
#Filterout unwanted disk types
no ide,xenblk,lsi_scsi,ahci,sd
no qed,qcow2v3,raw_dd,vmdk, usb2
no e1000-82540em,e1000-82545em,e1000-82544gc,xennet,nic_custom
only no_virtio_rng
only smp2
only no_9p_export
only no_pci_assignable
only (image_backend=filesystem)
only smallpages
smp = 32
vcpu_cores = 4
vcpu_threads = 8
vcpu_sockets = 1
# 32G
mem = 32768
flexible_nic_index = yes

variants:
    - guest_sanity:
        variants:
            - boundary:
                variants:
                    - network_max:
                        only qcow2
                        only scsi
                        variants:
                            - spaprvlan:
                                only spapr-vlan
                                nics += ' nic2 nic3 nic4 nic5 nic6 nic7 nic8 nic9 nic10 nic11 nic12 nic13 nic14 nic15 nic16'
                                only unattended_install.import.import.default_install.aio_native,remove_guest.without_disk
                            - virtionet:
                                only virtio_net
                                nic_model_nic1 = spapr-vlan
                                nics += ' nic2 nic3 nic4 nic5 nic6 nic7 nic8 nic9 nic10 nic11 nic12 nic13 nic14 nic15 nic16'
                                only unattended_install.import.import.default_install.aio_native,remove_guest.without_disk
```

xutian commented on virttest/virt_vm.py:
```diff
@@ -668,6 +670,19 @@ def get_address(self, index=0):
+                        try:
+                            mac = self.get_virsh_mac_address(i)
+                        except IndexError:
+                            break
```
why not convert  MAC address string to lowercase or uppercase  when right address_cache?

xutian replied:
Checked utils_env.py and 'utils_net' it looks we save MAC address in lowercase, I think "self.address_cache.get(mac.lower())" is enough.

sathnaga replied:
@xutian have fixed it

sathnaga commented:
@xutian @waynesun09 @will-Do could you please help merging this PR if no comments, Thanks in advance.

will-Do commented on virttest/virt_vm.py:
```diff
@@ -668,6 +668,17 @@ def get_address(self, index=0):
+                    # Try to poke for each index
+                    for i in range(len(macs)):
+                        try:
+                            mac = self.get_virsh_mac_address(i)
```
Here ```get_virsh_mac_address``` is a function in libvirt_vm.

sathnaga replied:
@will-Do it still works?

will-Do replied:
This will fail if the instance(self) is not libvirt_vm

will-Do commented:
We also met this problem, and I think maybe we can fix it in these specific cases by two ways:
1. When we try to login VM by a NIC without an IP address assigned to it, login by serial, then we could run ```dhclient ``` against the NIC inside the VM.
2. Or, we could try to login VM by all these NIC it has, just like the way in your patch, but in case level not the API level.
The reason here is that ```wait_for_login()```  always need a NIC(nic_index), if we add the new parameter try to login VM by all these NICs, which is  bit of violation to the original design of the API. 
What do you think?

sathnaga commented:
@will-Do 1. First approach of login to serial login if ssh failed should be enabled as a fall back in general. This can become permanent fix and will address the most of the failure cases.

2. I agree, it is bit of violation to original API, but changing wait_for_login() API will result in change in many places and break the flow, I do not want that to happen right now.

Due to this reason I introduced a new parameter which will enable poking all interfaces, we can consider this as a fixup till we get the stable/good fix considering the first approach?

sathnaga commented:
@will-Do ping

will-Do commented:
And for the second way I list above, I'm not mean to update wait_for_login, but update the part where calling it, handle all the NICs and exceptions.
I still think we should solve this problem case by case, no in the API level

sathnaga commented:
@will-Do one way is to give None as a index when we want flexible_index and keep this implementation,
in that case the it will not  violate the API and we can control from the calling function itself.
what say?

sathnaga commented:
@will-Do ping, can you comment on the above approach?

sathnaga commented:
@will-Do ping

balamuruhans commented on virttest/virt_vm.py:
```diff
@@ -668,6 +668,17 @@ def get_address(self, index=0):
+            if not arp_ip:
+                if (len(macs) > 1) and (flexible_index == "yes"):
+                    # Try to poke for each index
+                    for i in range(len(macs)):
```
it is good to have a name instead of i

balamuruhans commented on virttest/virt_vm.py:
```diff
@@ -676,6 +687,15 @@ def get_address(self, index=0):
                 ip_map = utils_net.parse_arp()
                 arp_ip = ip_map.get(nic.mac.lower())
+                if not arp_ip and (len(macs) > 1) and (flexible_index == "yes"):
+                    for i in range(len(macs)):
```
same here

will-Do commented on virttest/virt_vm.py:
```diff
@@ -638,6 +638,8 @@ def get_address(self, index=0):
                 be verified (using arping)
         """
+
+        flexible_index = self.params.get("flexible_nic_index", 'no')
```
@sathnaga 
According to this solution, my suggestion is make ```flexible_index``` as the 3rd param of this function:
```
     def get_address(self, index=0, flexible_index=False):
```
And use ```index=0``` as high priority

sathnaga replied:
@will-Do Have come up with the Version 2(https://github.com/avocado-framework/avocado-vt/issues/816) with all the changes implemented, hence closing this PR

sathnaga commented:
Created a V2 of this fix as referenced above.

sathnaga commented:
Created a V2 of this fix as referenced above.

sathnaga commented:
@will-Do I have reopened this initial version, with addressing comments, hope we can merge this one and
take further corrections if any as separate PRs. Thanks in advance.

sathnaga commented:
@will-Do @xutian Could you please help merge if no further comments, Thanks in advance :+1:

xutian commented:
LGTM

xutian commented on virttest/virt_vm.py:
```diff
@@ -668,6 +671,17 @@ def get_address(self, index=0):
+                    for idx in range(len(macs)):
+                        try:
+                            mac = self.get_virsh_mac_address(idx)
+                        except IndexError:
```
@sathnaga 
for tp-qemu test VM not create with libvirt, I guess self.get_virsh_mac_address() will raise TestFail exception, it will block tp-qemu  test.

sathnaga replied:
@xutian Thanks for the catch, am on it, will soon send a fix with that :-)

will-Do replied:
Yes, We talked about this problem before

will-Do commented on virttest/virt_vm.py:
```diff
@@ -638,6 +638,11 @@ def get_address(self, index=0):
+
+        flexible_index = ("yes" == self.params.get("flexible_nic_index", 'no'))
+        if flexible_index:
+            logging.debug("Incase of IP addr not found in index:%s,"
```
s/.../"In case index %s NIC has no IP address"

sathnaga commented:
@will-Do @xutian Have addressed the comments, have addressed the invalid address cache IP issue as well, hit into that scenario while testing.

sathnaga commented:
```
TESTS      : 4
 (1/4) guest_sanity.boundary.network_max.spaprvlan.qemu.qcow2.scsi.smp2.spapr-vlan.Fedora.24.ppc64le.io-github-autotest-qemu.unattended_install.import.import.default_install.aio_native: PASS (69.14 s)
 (2/4) guest_sanity.boundary.network_max.spaprvlan.qemu.qcow2.scsi.smp2.spapr-vlan.Fedora.24.ppc64le.io-github-autotest-libvirt.remove_guest.without_disk: PASS (1.83 s)
 (3/4) guest_sanity.boundary.network_max.virtionet.qemu.qcow2.scsi.smp2.virtio_net.Fedora.24.ppc64le.io-github-autotest-qemu.unattended_install.import.import.default_install.aio_native: PASS (178.84 s)
 (4/4) guest_sanity.boundary.network_max.virtionet.qemu.qcow2.scsi.smp2.virtio_net.Fedora.24.ppc64le.io-github-autotest-libvirt.remove_guest.without_disk: PASS (1.83 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0

```

sathnaga commented:
@xutian @will-Do Please help merge if no further comments, Thanks in advance :-)

xutian commented:
Merged,  @sathnaga thanks your contribution.

sathnaga commented:
@xutian Thanks :+1:
